src/cdk/testing/harness-environment.ts

```diff
@@ -147,43 +130,176 @@ export abstract class HarnessEnvironment<E> implements HarnessLoader, LocatorFac
    */
   protected abstract getAllRawElements(selector: string): Promise<E[]>;
 
-  private async _getAllHarnesses<T extends ComponentHarness>(
-      harnessType: ComponentHarnessConstructor<T> | HarnessPredicate<T>): Promise<T[]> {
-    const harnessPredicate = harnessType instanceof HarnessPredicate ?
-        harnessType : new HarnessPredicate(harnessType, {});
-    const elements = await this.getAllRawElements(harnessPredicate.getSelector());
-    return harnessPredicate.filter(elements.map(
-        element => this.createComponentHarness(harnessPredicate.harnessType, element)));
+  /** Gets a list of matching harnesses and test elements for a given query list. */
+  private async _getAllHarnessesAndTestElements<T extends (HarnessQuery<any> | string)[]>(
+      queries: T): Promise<LocatorFnResult<T>[]> {
+    const {allQueries, harnessQueries, elementQueries, harnessTypes} = _parseQueries(queries);
+    // Combine all of the queries into one large comma-delimited selector and use it to get all raw
+    // elements matching any of the individual queries.
+    const rawElements = await this.getAllRawElements(
+        [...elementQueries, ...harnessQueries.map(predicate => predicate.getSelector())].join(','));
+    // If there are only element queries, we can just return a `TestElement` for each raw element
+    // that was matched.
+    if (harnessQueries.length === 0) {
+      return this._matchRawElementsToTestElements(rawElements) as LocatorFnResult<T>[];
+    }
+    // If the queries contain only searches for a single harness type, we can just create instances
+    // of that class for all raw elements and filter out the ones that don't match any query.
+    if (elementQueries.length === 0 && harnessTypes.size === 1) {
+      return this._matchRawElementsToHarnessType(
+          rawElements, harnessQueries[0].harnessType, harnessQueries);
```

Comment:
I might be wrong here, but couldn't we remove that check and just handle it as part of `_matchRawElementsToHarnessesAndTestElements`? or will that return a different result if we have no element queries but only one harness type? (or is it just for performance to handle it separately?)

Reply:
Yeah, I did this just for performance reasons, to avoid unnecessary calls to `matchesSelector`. Though looking now at how this code came out, it shouldn't be too bad to fold it all into the same code path.